src/Microsoft.Azure.SignalR.Common/ServiceConnections/ServiceConnectionContainerBase.cs

```diff
@@ -598,14 +607,15 @@ public void Stop()
                     }
                     if (Interlocked.Decrement(ref _counter) == 0)
                     {
-                        _timer.Stop();
+                        _timer?.Stop();
+                        _timer = null;
                     }
                 }
             }
 
             public void Dispose()
             {
-                _timer.Stop();
+                _timer?.Stop();
```

Comment:
also set null after stop?

Reply:
if we null it out here, then we should keep the null check in Stop() to avoid race condition between Dispose and Stop, right?

Reply:
I'm wondering whether the race condition between Dispose and Stop could happen. Just notice TimeAwaitable has a `Dispose` method, or shall we call `_timer.Dispose()` here?

Reply:
From looking at the `TimerAwaitable` implementation, it seems the proper way is to call `_timer.Stop()` and only then dispose it. The `using(timer)` I added inside `PingAsync` does just that. Thus calling `_timer.Stop` in `CustomizedPingTimer.Dispose` is following the same pattern. Although if nobody calls `CustomizedPingTimer.Start` then we'd never call `_timer.Dispose`... And calling Dispose more than once doesn't have any adverse effects. However calling `Dispose` and then `Stop` on the timer does throw...

Which brings me to the race between `Dispose` and then `Stop`. I am also not 100% sure what scenario will cause this. As far as I am aware nobody right now is ever disposing `MultiEndpointServiceConnectionContainer` instances (this is the only other route of getting the containers disposed). But it seems like a good practice to make Dispose as safe to call as possible.

Reply:
> I'm wondering whether the race condition between Dispose and Stop could happen. Just notice TimeAwaitable has a `Dispose` method, or shall we call `_timer.Dispose()` here?

BTW, it is explicitly implemented as `void IDisposable.Dispose()` which implies that "nobody _should_ ever call it, other than from using()". From further looking into its code it seems that disposing it is not necessary if nobody called `Start`.

Reply:
Yes, but currently we split the Start/Stop operation and not able to handle cleanup by a single using. How about call `((IDisposable)_timer).Dispose();`?